test/integrationTests/integrationHelpers.ts

```diff
@@ -15,19 +17,41 @@ export async function activateCSharpExtension(): Promise<void> {
         vscode.extensions.getExtension<CSharpExtensionExports>(vscodeDotnetRuntimeExtensionId);
     if (!dotnetRuntimeExtension) {
         await vscode.commands.executeCommand('workbench.extensions.installExtension', vscodeDotnetRuntimeExtensionId);
-        await reloadWindow();
+        await vscode.commands.executeCommand('workbench.action.reloadWindow');
     }
 
     const csharpExtension = vscode.extensions.getExtension<CSharpExtensionExports>('ms-dotnettools.csharp');
     if (!csharpExtension) {
         throw new Error('Failed to find installation of ms-dotnettools.csharp');
     }
 
+    // Run a restore manually to make sure the project is up to date since we don't have automatic restore.
+    await testAssetWorkspace.restoreLspToolsHostAsync();
+
+    // If the extension is already active, we need to restart it to ensure we start with a clean server state.
+    // For example, a previous test may have changed configs, deleted restored packages or made other changes that would put it in an invalid state.
+    let shouldRestart = false;
```

Comment:
@jasonmalinowski could use your thoughts here on how to better handle this.  We have a couple issues here
1.  The server might have activated before we restored the project.  This might be fixed if we auto-restore project when we load it.
2.  A previous test run might have cleaned the workspace (including restored assets), leaving the server missing references.  We do run restore here again, but we currently have no way of waiting for the server to have actually updated (other than explicitly restarting it).

We could fire events every time the server reloads a project?  But it is hard to tell when that is 'done'

Reply:
Should we do something similar to our asynchronous listeners like we do in Visual Studio for Windows? We could introduce a test-only LSP endpoint that is fired off and completed only once all the work is done. We could either make it a one-off for the project system, or actually wire it into our async listener API so we could use it generally (and pass along the feature name).

Reply:
Yup I like this idea - basically send the `waitForAllOperations` request to the server and have it return once its done.  I'll take a look at doing this in a followup (this ones already huge)